File: bin/render_vrml/render_board.js
/* 
  ############################################################################################
  # Hellen-One: A board rendering server script.
  # (c) andreika <prometheus.pcb@gmail.com>
  ############################################################################################
  
  Node.js is required to run this script: https://nodejs.org/en/download/
  Also please install PNGJS Node modules before running:
     > npm install --prefix ./bin/render_vrml pngjs
*/

function getPixel(img, x, y) {
    if (x < 0 || y < 0 || x >= img.width || y >= img.height)
    	return [0, 0, 0, 0];
	var idx = (img.width * y + x) << 2;
	return [ img.data[idx], img.data[idx + 1], img.data[idx + 2], img.data[idx + 3] ];
}

function createBoardImg(pcbImg, outlineImg, compImg, compImgOffset) {
	var boardImg = new PNG({ 
		width: Math.max(pcbImg.width, outlineImg.width, compImg.width), 
		height: Math.max(pcbImg.height, outlineImg.height, compImg.height) });
	var pcbOffY = (pcbImg.height < outlineImg.height) ? -(outlineImg.height - pcbImg.height) : 0;
	// Blit the pcbImg using the outlineImg mask and add compImg.
	// We cannot use PNG.bitblt() for that
	for (var y = 0; y < boardImg.height; y++) {
    	for (var x = 0; x < boardImg.width; x++) {
        	var bPixel = getPixel(pcbImg, x, y + pcbOffY);
            var cPixel = getPixel(compImg, x + compImgOffset[0], y + compImgOffset[1]);
        	var a = parseFloat(cPixel[3]) / 255.0;
        	var na = 1.0 - a;
        	var idx = (boardImg.width * y + x) << 2;
        	boardImg.data[idx + 0] = na * bPixel[0] + a * cPixel[0];
        	boardImg.data[idx + 1] = na * bPixel[1] + a * cPixel[1];
        	boardImg.data[idx + 2] = na * bPixel[2] + a * cPixel[2];
        	boardImg.data[idx + 3] = na * getPixel(outlineImg, x, y)[0] + a * 255;
		}
	}
	return boardImg;
}

try {
	var PNG = require("pngjs").PNG;
	// built-in modules
	var fs = require("fs");

	var args = process.argv.slice(2);
	if (args.length < 4) {
		console.log("* Error! Please specify correct arguments to run this script!");
		process.exit(1)
	}
	// arguments
	var pcbImgFile = args[0];
	var outlineImgFile = args[1];
	var compImgFile = args[2];
	var boardImgFile = args[3];
	var compImgOffset = args[4].split(",").map((x) => parseInt(x));
	
	console.log("* Reading the pcb image...");
	var pcbImg = PNG.sync.read(fs.readFileSync(pcbImgFile));
  	console.log("* Reading the components image with offset (" + compImgOffset[0] + "," + compImgOffset[1] + ")...");
	var compImg = PNG.sync.read(fs.readFileSync(compImgFile));
  	console.log("* Reading the outline image...");
	var outlineImg = PNG.sync.read(fs.readFileSync(outlineImgFile));

	console.log("* Creating the final board image...");
	var boardImg = createBoardImg(pcbImg, outlineImg, compImg, compImgOffset);
  	console.log("* Saving the board image...");
	fs.writeFileSync(boardImgFile, PNG.sync.write(boardImg, { colorType: 6 }));

  	console.log("* Done! Exiting...");

} catch(e) {
    if (e.message.indexOf("Cannot find module") !== -1) {
    	console.error("Error! `pngjs` library is not installed? Try running `npm install --prefix ./bin/render_vrml`.");
    }
    console.log(e);
    process.exit(e.code);
}

File: bin/render_vrml/render_board.py
#!/usr/bin/env python
############################################################################################
# Hellen-One: A board rendering server script.
# Python 3.5+ is required.
# Dependencies: Pillow
# (c) andreika <prometheus.pcb@gmail.com>
############################################################################################

import os, sys
from PIL import Image

if (len(sys.argv) < 6):
	print ("* Error! Please specify correct arguments to run this script!")
	sys.exit(1)

pcbImgFile = sys.argv[1]
outlineImgFile = sys.argv[2]
compImgFile = sys.argv[3]
boardImgFile = sys.argv[4]
compImgOffset = [int(n) for n in sys.argv[5].split(",")]

class ImageObject:
	width = 0
	height = 0
	data = []

def getPixel(img, x, y):
	if (x < 0 or y < 0 or x >= img.width or y >= img.height):
		return [0, 0, 0, 0]
	return img.data[x, y]

def createBoardImg(pcbImg, outlineImg, compImg, compImgOffset):
	width = max([pcbImg.width, outlineImg.width, compImg.width])
	height = max([pcbImg.height, outlineImg.height, compImg.height])
	boardImg = Image.new('RGBA', (width, height))

	pcbOffY = -(outlineImg.height - pcbImg.height) if (pcbImg.height < outlineImg.height) else 0
	# Blit the pcbImg using the outlineImg mask and add compImg.
	# We cannot use PNG.bitblt() for that
	for y in range(0, boardImg.height):
		for x in range(0, boardImg.width):
			bPixel = getPixel(pcbImg, x, y + pcbOffY)
			cPixel = getPixel(compImg, x + compImgOffset[0], y + compImgOffset[1])
			a = float(cPixel[3]) / 255.0
			na = 1.0 - a
			pr = int(na * bPixel[0] + a * cPixel[0])
			pg = int(na * bPixel[1] + a * cPixel[1])
			pb = int(na * bPixel[2] + a * cPixel[2])
			pa = int(na * getPixel(outlineImg, x, y)[0] + a * 255)
			boardImg.putpixel((x, y), (pr, pg, pb, pa))
	return boardImg

def loadImage(fileName):
	pimg = Image.open(fileName).convert("RGBA")
	img = ImageObject()
	img.data = pimg.load()
	img.width = pimg.size[0]
	img.height = pimg.size[1]
	return img

print ("* Reading the pcb image...")
pcbImg = loadImage(pcbImgFile)
print ("* Reading the components image with offset (" + str(compImgOffset[0]) + "," + str(compImgOffset[1]) + ")...")
compImg = loadImage(compImgFile)
print ("* Reading the outline image...")
outlineImg = loadImage(outlineImgFile)

print ("* Creating the final board image...")
boardImg = createBoardImg(pcbImg, outlineImg, compImg, compImgOffset)
print ("* Saving the board image...")
boardImg.save(boardImgFile)

print ("* Done! Exiting...")
